Externalise a few timeouts & fix timeout for hostSupportsUefi in libvirt ready command wrapper

vishesh92 opened the pull request:
### Description

This PR fixes bug introduced in #8502. Timeout for script execution was set to 60 ms instead of 60s which resulted in host not getting UEFI enabled. This is a blocker for 4.19 release.

We do this by introducing a new agent parameter `agent.script.timeout` (default - 60 seconds) to use as a timeout for the script checking host's UEFI status.

We also externalize the timeout for the ReadyCommand by introducing a new global setting `ready.command.wait` (default - 60 seconds).

For ModifyStoragePoolCommand, we don't externalize the timeout to avoid confusion for the user. Since, the required timeout can vary depending on the provider in use and we are only setting the wait for default host listener for now. Instead, we reuse the global `wait` setting by dividing it by `5` making the default value of 6 minutes (1800/5 = 360s) for ModifyStoragePoolCommand.

Note: the actual time, the MS waits is twice the wait set for a Command. Check reference code below.
https://github.com/apache/cloudstack/blob/19250403e645c76f60b17aa4aeb4dc915f5ca206/engine/orchestration/src/main/java/com/cloud/agent/manager/AgentAttache.java#L406-L442

<!--- Describe your changes in DETAIL - And how has behaviour functionally changed. -->

<!-- For new features, provide link to FS, dev ML discussion etc. -->
<!-- In case of bug fix, the expected and actual behaviours, steps to reproduce. -->

<!-- When "Fixes: #<id>" is specified, the issue/PR will automatically be closed when this PR gets merged -->
<!-- For addressing multiple issues/PRs, use multiple "Fixes: #<id>" -->
<!-- Fixes: # -->

<!--- ******************************************************************************* -->
<!--- NOTE: AUTOMATION USES THE DESCRIPTIONS TO SET LABELS AND PRODUCE DOCUMENTATION. -->
<!--- PLEASE PUT AN 'X' in only **ONE** box -->
<!--- ******************************************************************************* -->

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [x] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial


### Screenshots (if appropriate):


### How Has This Been Tested?

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->

#### How did you try to break this feature and the system with this change?

<!-- see how your change affects other areas of the code, etc. -->


<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/main/CONTRIBUTING.md) document -->

vishesh92 commented:
@blueorangutan package

shwstppr commented:
@blueorangutan package

blueorangutan commented:
@shwstppr a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

sureshanaparti reviewed (approved):
code lgtm

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 8412

DaanHoogland reviewed (approved):
clgtm

DaanHoogland commented:
is this maybe suitable for 4.18.2 even?

vishesh92 commented:
> is this maybe suitable for 4.18.2 even?

We will need to add changes in this PR https://github.com/apache/cloudstack/pull/8502 to 4.18 as well.

weizhouapache reviewed (approved):
code lgtm

vishesh92 commented:
@blueorangutan package

vishesh92 commented:
@blueorangutan package

weizhouapache commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtReadyCommandWrapper.java:
```diff
@@ -51,11 +53,12 @@ public Answer execute(final ReadyCommand command, final LibvirtComputingResource
         if (isUbuntuHost) {
             cmd = "dpkg -l ovmf";
         }
         s_logger.debug("Running command : " + cmd);
```
add timeout value to debug message ?

sureshanaparti commented on engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/provider/DefaultHostListener.java:
```diff
@@ -55,8 +57,12 @@
     private static final Logger s_logger = Logger.getLogger(DefaultHostListener.class);
+    ConfigKey<Integer> ModifyStoragePoolCommandWait = new ConfigKey<Integer>("Advanced", Integer.class,
+            "modify.storage.pool.command.wait", "60",
+            "Time in seconds to wait for ModifyStoragePoolCommand command to return", true);
```
CS have lot of such cmds, maybe not good idea to add a config at cmd level. (later, might end up one config for each cmd). generic configs, eg. storage.command.wait / host.command.wait, based on the category with appropriate descriptions might be good.

vishesh92 replied:
I agree with you. But generic config might not work. e.g. `storage.command.wait` as 60 seconds will work for `ModifyStoragePoolCommand` but not might be appropriate for other storage related commands. It needs to be specific to a `Command`.

IMO, it would be better to have something which allows the user to set configuration if required like `command.<command name>.wait` and wait is set only if the value is set. If it's not set, then some default value is used.

weizhouapache replied:
What about
agent.startup.command.timeout
agent.startup.script.timeout 
?

vishesh92 commented:
@blueorangutan package

vishesh92 commented:
@blueorangutan test

vishesh92 commented:
@blueorangutan test matrix

blueorangutan commented:
@vishesh92 a [SL] Trillian-Jenkins matrix job (centos7 mgmt + xenserver71, rocky8 mgmt + vmware67u3, centos7 mgmt + kvmcentos7) has been kicked to run smoke tests

vishesh92 commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtReadyCommandWrapper.java:
```diff
@@ -51,11 +53,12 @@ public Answer execute(final ReadyCommand command, final LibvirtComputingResource
 
     private boolean hostSupportsUefi(boolean isUbuntuHost) {
         String cmd = "rpm -qa | grep -i ovmf";
+        int timeout = AgentPropertiesFileHandler.getPropertyValue(AgentProperties.LIBVIRT_HOST_UEFI_SCRIPT_TIMEOUT) * 1000; // Get property value & convert to milliseconds
```
Add a common timeout for all scripts here.

vishesh92 commented on engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/provider/DefaultHostListener.java:
```diff
@@ -121,7 +135,7 @@ private NicTO createNicTOFromNetworkAndOffering(NetworkVO networkVO, NetworkOffe
         StoragePool pool = (StoragePool) this.dataStoreMgr.getDataStore(poolId, DataStoreRole.Primary);
         ModifyStoragePoolCommand cmd = new ModifyStoragePoolCommand(true, pool);
-        cmd.setWait(60);
+        cmd.setWait(ModifyStoragePoolCommandWait.value());
```
Use (wait.timeout / 6) here.

blueorangutan commented:
<b>[SF] Trillian test result (tid-8925)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 47048 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8547-t8925-kvm-centos7.zip
Smoke tests completed. 121 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

vishesh92 commented:
@blueorangutan package

blueorangutan commented:
@vishesh92 a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
<b>[SF] Trillian test result (tid-8924)</b>
Environment: vmware-67u3 (x2), Advanced Networking with Mgmt server r8
Total time taken: 48332 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8547-t8924-vmware-67u3.zip
Smoke tests completed. 120 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_balanced_drs_algorithm | `Failure` | 135.16 | test_cluster_drs.py

shwstppr commented on engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/provider/DefaultHostListener.java:
```diff
@@ -121,7 +122,7 @@ private NicTO createNicTOFromNetworkAndOffering(NetworkVO networkVO, NetworkOffe
         StoragePool pool = (StoragePool) this.dataStoreMgr.getDataStore(poolId, DataStoreRole.Primary);
         ModifyStoragePoolCommand cmd = new ModifyStoragePoolCommand(true, pool);
-        cmd.setWait(60);
+        cmd.setWait(Wait.value()/5);
```
@vishesh92 do we need to add some logging here?

sureshanaparti replied:
```suggestion
        cmd.setWait(Wait.value() / 5);
```

sureshanaparti commented on engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/provider/DefaultHostListener.java:
```diff
@@ -121,7 +122,9 @@ private NicTO createNicTOFromNetworkAndOffering(NetworkVO networkVO, NetworkOffe
-        cmd.setWait(60);
+        cmd.setWait(Wait.value() / 5);
+        s_logger.debug(String.format("Sending modify storage pool command to agent: %d for storage pool: %d with timeout %d seconds",
+                hostId, poolId, cmd.getWait() / 5));
```
```suggestion
                hostId, poolId, cmd.getWait()));
```

harikrishna-patnala commented on agent/src/main/java/com/cloud/agent/properties/AgentProperties.java:
```diff
@@ -659,6 +659,14 @@ public class AgentProperties{
+     * Data type: Integer.<br>
+     * Default value: <code>60</code>
+     */
+    public static final Property<Integer> AGENT_SCRIPT_TIMEOUT = new Property<>("agent.script.timeout", 60);
```
@vishesh92 we need to add this in agent.properties to make it configurable

vishesh92 replied:
We don't need to. If the user adds it in the file, it will get picked up.

harikrishna-patnala replied:
yeah correct but that has to be known to the operator right, we normally put them in the properties file and comment them 

eg.

# Instance Conversion from Vmware to KVM through virt-v2v. Enable verbose mode
# virtv2v.verbose.enabled=false

vishesh92 commented:
@blueorangutan package

blueorangutan commented:
@vishesh92 a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 8435

vishesh92 commented:
@blueorangutan test matrix

blueorangutan commented:
@vishesh92 a [SL] Trillian-Jenkins matrix job (centos7 mgmt + xenserver71, rocky8 mgmt + vmware67u3, centos7 mgmt + kvmcentos7) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-8936)</b>
Environment: xenserver-71 (x2), Advanced Networking with Mgmt server 7
Total time taken: 41623 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8547-t8936-xenserver-71.zip
Smoke tests completed. 120 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_trigger_shutdown | `Failure` | 341.90 | test_safe_shutdown.py

vishesh92 commented on agent/src/main/java/com/cloud/agent/properties/AgentProperties.java:
```diff
@@ -659,6 +659,14 @@ public class AgentProperties{
 
+    /**
+     * Time (in seconds) to wait for scripts to complete.<br>
+     * This is not used for every script execution.<br>
```
```suggestion
     * This is currently used only while checking if the host supports UEFI.<br>
```

vishesh92 commented on agent/conf/agent.properties:
```diff
@@ -378,6 +378,10 @@ iscsi.session.cleanup.enabled=false
 #stop.script.timeout=120
 
+# Time (in seconds) to wait for scripts to complete.
+# This is not used for every script execution.
```
```suggestion
# This is currently used only while checking if the host supports UEFI.
```

harikrishna-patnala reviewed (approved):
code LGTM

weizhouapache reviewed (approved):
code lgtm

weizhouapache commented:
@blueorangutan test ubuntu22 kvm-ubuntu22

blueorangutan commented:
@weizhouapache a [SL] Trillian-Jenkins test job (ubuntu22 mgmt + kvm-ubuntu22) has been kicked to run smoke tests

weizhouapache commented:
@blueorangutan test alma9 kvm-alma9

blueorangutan commented:
<b>[SF] Trillian test result (tid-8938)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 52708 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8547-t8938-kvm-centos7.zip
Smoke tests completed. 120 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_08_upgrade_kubernetes_ha_cluster | `Failure` | 940.82 | test_kubernetes_clusters.py

blueorangutan commented:
<b>[SF] Trillian test result (tid-8937)</b>
Environment: vmware-67u3 (x2), Advanced Networking with Mgmt server r8
Total time taken: 53125 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8547-t8937-vmware-67u3.zip
Smoke tests completed. 120 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_list_vms_metrics_history | `Error` | 15.12 | test_metrics_api.py

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

GutoVeronezi commented:
-1 (binding)

@vishesh92 this PR does not fix #8506. The issue and the thread it mentions is related to a generic solution, which was not even discussed.

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 8440

shwstppr commented:
> -1 (binding)
> 
> @vishesh92 this PR does not fix #8506. The issue and the thread it mentions is related to a generic solution, which was not even discussed.

@GutoVeronezi would it make sense to keep the issue open for further work and continue with this PR or are you -1 on the PR changes as well?

DaanHoogland commented:
> -1 (binding)

binding does not apply to PRs. That said, if you have objections and a way forward we will not merge (on any solidly argued -1)

> @vishesh92 this PR does not fix #8506. The issue and the thread it mentions is related to a generic solution, which was not even discussed.

@GutoVeronezi , I suppose you are referring to the solution mentioned in this comment: https://github.com/apache/cloudstack/pull/8502#discussion_r1450506118 ?

If so, this is a good idea and would obsolete this PR, but not without implementation. Do you have code ready to replace the code here?

vishesh92 commented:
> -1 (binding)
> 
> @vishesh92 this PR does not fix #8506. The issue and the thread it mentions is related to a generic solution, which was not even discussed.

@GutoVeronezi I agree that this PR does not fix #8506. We need to revisit the timeouts for Commands on both the MS (#8506) & agents (#8011 - Agents don't even consider timeout which is set in the Commands).

The main purpose of #8502 was to fix the failures we were seeing in e2e tests, but I introduced another bug (because of using 60ms instead of 60s for the script's timeout).

In this PR, I wanted to fix the bug I introduced and address your concerns at least for the changes I did in my previous PR (#8502).

Regarding externalizing timeouts for all commands, IMO we shouldn't try to fix it for now. I would like to do that but it would be a very big change and it will require more discussion and planning because the same `Command` can processed by different providers depending on an environment and the required timeout can be different depending on the scenario and the setup.

I would like to get this PR merged for now since it's a blocker for unblock 4.19 release. Let me know what you think?

GutoVeronezi commented:
> @GutoVeronezi would it make sense to keep the issue open for further work...

Sure do, because the issue is not being fixed by this PR.

> ... are you -1 on the PR changes as well?

No. Indeed, I proposed the externalization of the timeout settings, vide https://github.com/apache/cloudstack/pull/8502#discussion_r1450410628 (and the author of the PR was against it at first: https://github.com/apache/cloudstack/pull/8502#discussion_r1452036208). My point is that this PR does not even touch the essence of the issue; therefore, saying it solves the issue does not make sense.

> binding does not apply to PRs.

Thanks, @DaanHoogland. I think I missed this detail from the docs.

> @GutoVeronezi , I suppose you are referring to the solution mentioned in this comment: [#8502 (comment)](https://github.com/apache/cloudstack/pull/8502#discussion_r1450506118) ?
> 
> If so, this is a good idea and would obsolete this PR, but not without implementation. Do you have code ready to replace the code here?

No, there is no code ready yet; before implementing something every each way, we should discuss the pros and cons. That is the purpose of the issue.

GutoVeronezi commented on engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/provider/DefaultHostListener.java:
```diff
@@ -121,7 +122,9 @@ private NicTO createNicTOFromNetworkAndOffering(NetworkVO networkVO, NetworkOffe
         StoragePool pool = (StoragePool) this.dataStoreMgr.getDataStore(poolId, DataStoreRole.Primary);
         ModifyStoragePoolCommand cmd = new ModifyStoragePoolCommand(true, pool);
-        cmd.setWait(60);
+        cmd.setWait(Wait.value() / 5);
```
Why divide by 5?

vishesh92 replied:
For ModifyStoragePoolCommand, we don't externalize the timeout to avoid confusion for the user. Since, the required timeout can vary depending on the provider in use and we are only setting the wait for default host listener for now. Instead, we reuse the global wait setting by dividing it by 5 making the default value of 6 minutes (1800/5 = 360s) for ModifyStoragePoolCommand.

GutoVeronezi replied:
It still does not make sense. Why 5? Just to match 6 minutes when the default `wait` is 1800?

vishesh92 replied:
Yes. Reasoning behind this is that user might decrease the wait in some scenario. It's quite unlikely but if the user sets it to 5 min, we will wait at least 1 minute here.

P.S. - the actual wait time is twice the value set for wait. Check the note in the description.

GutoVeronezi replied:
User also might increase `wait` to 18000 and we would have 1 hour here. It does not make sense to use an unfounded math function to derive a value; please, either use the same configuration or externalize a new one.

vishesh92 replied:
It's more about the worst case scenario. It will wait for the specified time only if the agent is stuck because of some reason. Before #8502, if the wait was set to 18000, this command would have waited for 10 hours (5 hours *2). This solution doesn't completely solve the the problem but reduces the issues to some extent until issues #8506 or #8011 is resolved.

GutoVeronezi replied:
In this case, externalizing a new configuration is the best choice.

GutoVeronezi replied:
In contrast with taking a value and dividing it by a random number, externalizing a new configuration would be the best option. We will not externalize every command timeout right now, so it would not be a problem. I agree a more robust solution, like what I mentioned in https://github.com/apache/cloudstack/pull/8502#discussion_r1450506118, is better; however, we have to discuss it first.

As last resort, if you do not want to externalize a new configuration, creating a constant with javadoc for this case is better than deriving the value.

GutoVeronezi commented:
> I would like to get this PR merged for now since it's a blocker for unblock 4.19 release. Let me know what you think?

I see #8502 was a blocker, but why this one is a blocker?

GutoVeronezi commented:
@vishesh92, please, improve the description of the PR to reflect the real proposal (which is different from issue #8506), mentioning PR #8502 and explaining why it is a blocker.

vishesh92 commented:
> I see #8502 was a blocker, but why this one is a blocker?
I used 60 ms instead of 60000 ms as timeout for the script. This resulted in all hosts having UEFI status as false.

> @vishesh92, please, improve the description of the PR to reflect the real proposal (which is different from issue #8506), mentioning PR #8502 and explaining why it is a blocker.

I have updated the PR description.

vishesh92 commented:
@blueorangutan package

blueorangutan commented:
@vishesh92 a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 8443

vishesh92 commented:
@blueorangutan test matrix

blueorangutan commented:
@vishesh92 a [SL] Trillian-Jenkins matrix job (centos7 mgmt + xenserver71, rocky8 mgmt + vmware67u3, centos7 mgmt + kvmcentos7) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-8941)</b>
Environment: kvm-ubuntu22 (x2), Advanced Networking with Mgmt server u22
Total time taken: 50172 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8547-t8941-kvm-ubuntu22.zip
Smoke tests completed. 121 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

blueorangutan commented:
<b>[SF] Trillian test result (tid-8942)</b>
Environment: kvm-alma9 (x2), Advanced Networking with Mgmt server a9
Total time taken: 55621 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8547-t8942-kvm-alma9.zip
Smoke tests completed. 121 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

blueorangutan commented:
<b>[SF] Trillian test result (tid-8946)</b>
Environment: xenserver-71 (x2), Advanced Networking with Mgmt server 7
Total time taken: 43130 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8547-t8946-xenserver-71.zip
Smoke tests completed. 121 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

blueorangutan commented:
<b>[SF] Trillian test result (tid-8948)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 43265 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8547-t8948-kvm-centos7.zip
Smoke tests completed. 121 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

blueorangutan commented:
<b>[SF] Trillian test result (tid-8947)</b>
Environment: vmware-67u3 (x2), Advanced Networking with Mgmt server r8
Total time taken: 48946 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8547-t8947-vmware-67u3.zip
Smoke tests completed. 119 look OK, 2 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
ContextSuite context=TestAccounts>:setup | `Error` | 0.00 | test_accounts.py
ContextSuite context=TestAddVmToSubDomain>:setup | `Error` | 0.00 | test_accounts.py
test_DeleteDomain | `Error` | 9.30 | test_accounts.py
test_forceDeleteDomain | `Failure` | 9.18 | test_accounts.py
ContextSuite context=TestRemoveUserFromAccount>:setup | `Error` | 10.38 | test_accounts.py
ContextSuite context=TestTemplateHierarchy>:setup | `Error` | 12.21 | test_accounts.py
test_02_balanced_drs_algorithm | `Error` | 429.31 | test_cluster_drs.py

shwstppr reviewed (approved):
LGTM, uefi cmd running fine,

```
[root@pr8547-t6913-kvm-rocky8-kvm1 ~]# systemctl restart cloudstack-agent && tail -f /var/log/cloudstack/agent/agent.log 
2024-01-26 11:04:17,735 DEBUG [resource.wrapper.LibvirtReadyCommandWrapper] (agentRequest-Handler-4:null) (logid:c84198da) Running command : [rpm -qa | grep -i ovmf] with timeout : 60000 ms
2024-01-26 11:04:17,735 DEBUG [utils.script.Script] (agentRequest-Handler-4:null) (logid:c84198da) Executing command [/bin/bash -c rpm -qa | grep -i ovmf ].
2024-01-26 11:04:18,028 DEBUG [utils.script.Script] (agentRequest-Handler-4:null) (logid:c84198da) Successfully executed process [22872] for command [/bin/bash -c rpm -qa | grep -i ovmf ].
```

Without fix,
```
[root@ref-trl-6184-k-Mr8-nicolas-vazquez-kvm2 ~]# tail -f /var/log/cloudstack/agent/agent.log
2024-01-23 04:53:27,983 INFO  [cloud.agent.Agent] (agentRequest-Handler-5:null) (logid:b473a821) Set agent id 7
2024-01-23 04:53:27,984 INFO  [cloud.agent.Agent] (agentRequest-Handler-5:null) (logid:b473a821) Ready command is processed for agent id = 7
2024-01-23 04:53:28,064 WARN  [utils.script.Script] (Script-4:null) (logid:) Interrupting script.
2024-01-23 04:53:28,065 WARN  [utils.script.Script] (agentRequest-Handler-5:null) (logid:b473a821) Process [320455] for command [/bin/bash -c rpm -qa | grep -i ovmf ] timed out. Output is [].
```

shwstppr commented:
@GutoVeronezi are your concerns with this PR addressed? I'll proceed with 4.19RC4 if no objections
